WasatchNET/MockSpectrometer.cs

```diff
@@ -0,0 +1,546 @@
+
+                }
+                else
+                {
+                    double[] minSpectrum;
+                    double[] maxSpectrum;
+                    int minIntegration;
+                    int maxIntegration;
+
+                    if (index == -1)
+                    {
+                        var enumerator = spectra.GetEnumerator();
+                        enumerator.MoveNext();
+                        minSpectrum = enumerator.Current.Value;
+                        minIntegration = enumerator.Current.Key;
+                        enumerator.MoveNext();
+
+                        maxSpectrum = enumerator.Current.Value;
+                        maxIntegration = enumerator.Current.Key;
+                    }
+                    else
+                    {
+                        var enumerator = spectra.GetEnumerator();
+                        for (int i = 0; i < spectra.Count - 1; ++i)
+                            enumerator.MoveNext();
+                        minSpectrum = enumerator.Current.Value;
+                        minIntegration = enumerator.Current.Key;
+                        enumerator.MoveNext();
+
+                        maxSpectrum = enumerator.Current.Value;
+                        maxIntegration = enumerator.Current.Key;
+                    }
+
+                    for (int i = 0; i < pixels; ++i)
+                    {
+                        double slope = (maxSpectrum[i] - minSpectrum[i]) / (maxIntegration - minIntegration);
+                        double delta = slope * (integrationTimeMS - maxIntegration);
+                        final[i] = maxSpectrum[i] + delta;
+                    }
+
+
+                }
+            }
+
+            if (sampleMethod == SAMPLE_METHOD.NOISY_LINEAR_INTERPOLATION)
+                final = addNoise(final, 20, 1);
+
+            return final;
+        }
+
+        double addNoise(double data, double mean, double sd)
+        {
+            Random rand = new Random(); //reuse this if you are generating many
+            double u1 = 1.0 - rand.NextDouble(); //uniform(0,1] random doubles
+            double u2 = 1.0 - rand.NextDouble();
+            double randStdNormal = Math.Sqrt(-2.0 * Math.Log(u1)) *
+                         Math.Sin(2.0 * Math.PI * u2); //random normal(0,1)
+            double randNormal = mean + sd * randStdNormal;
+
+            double noised = (double)data + randNormal;
+
+            return noised;
+        }
+
+        double[] addNoise(double[] data, double mean, double sd)
+        {
+            Random rand = new Random(); //reuse this if you are generating many
+
+            double[] noised = new double[data.Length];
+
+            int index = 0;
+            foreach (double d in data)
+            {
+                double u1 = 1.0 - rand.NextDouble(); //uniform(0,1] random doubles
+                double u2 = 1.0 - rand.NextDouble();
+                double randStdNormal = Math.Sqrt(-2.0 * Math.Log(u1)) *

```

Comment:
It is a long time since I've seen someone go to that much trouble to make noise...